### PR TITLE
deprecates babel/polyfill in favor of core-js@3

### PR DESCRIPTION
**Pull request by somebody32**
[babel/polyfill is deprecated](https://github.com/zloirock/core-js/blob/master/docs/2019-03-19-core-js-3-babel-and-a-look-into-the-future.md) and using core-js directly is the way now.

This pr adds the required modifications to support core-js@3. 
~I didn't remove `babel/polyfill` from the dependencies to not make this change breaking.~

Update: removed `babel/polyfill` because updating `core-js` would break old usage anyway, so updating after that PR would require changes to the user's codebase

### CODE REVIEW

File: package/rules/node_modules.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ const { nodeEnv } = require('../env')
 module.exports = {
   test: /\.(js|mjs)$/,
   include: /node_modules/,
-  exclude: /@babel(?:\/|\\{1,2})runtime/,
+  exclude: /@babel(?:\/|\\{1,2})runtime|core-js/,
```
Comment (somebody32) [inline package/rules/node_modules.js]: running babel on core-js is causing [issues](https://github.com/zloirock/core-js/issues/514)

**Comment by renchap** [comment]
https://github.com/rails/webpacker/pull/2031#discussion_r269065961 : this is important, as this breaks IE11 for any person polyfilling `Symbol` and using `core-js@3`.

**Comment by jakeNiemiec** [comment]
Babel PR: https://github.com/babel/babel/pull/7646

File: package.json [hunk 2]
```diff
@@ -12,21 +12,22 @@
...
-    "@babel/plugin-transform-destructuring": "^7.3.2",
-    "@babel/plugin-transform-regenerator": "^7.3.4",
-    "@babel/plugin-transform-runtime": "^7.3.4",
-    "@babel/polyfill": "^7.2.5",
-    "@babel/preset-env": "^7.3.4",
-    "@babel/runtime": "^7.3.4",
+    "@babel/plugin-transform-destructuring": "^7.4.0",
+    "@babel/plugin-transform-regenerator": "^7.4.0",
+    "@babel/plugin-transform-runtime": "^7.4.0",
+    "@babel/polyfill": "^7.4.0",
```
[the remark this answers is not in the document]
Comment (somebody32) [reply package.json]: sorry, needed what? `@babel/polyfill`? yes, if we remove it, then it is going to be a breaking change

**Comment by connorshea** [comment]
We should maybe have a changelog entry written up that has an upgrade guide.

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicolo-ribaudo) [reply package.json]: (Disclaimer: I know Babel and core-js pretty well, but not webpacker)

Why would it be a breaking change? Is it exposed to your users or is it only added by webpacker?

[the remark this answers is not in the document]
Comment (jakeNiemiec) [reply package.json]: image
@somebody32 The document you linked seems to indicate that it needs to be removed. Apart from that, I feel like this is defenitly a breaking change since users will need to change their pack files.

[the remark this answers is not in the document]
Comment (somebody32) [reply package.json]: @nicolo-ribaudo @jakeNiemiec you're right, I for some reason thought that core-js@3 and @babel/polyfill could coexist for some time, but looks like it is not the cause. My idea initially was not to remove @babel/polyfill, so builds which indirectly dependant on it will not be broken, but looks like it is safer to remove it so there will be no strange compatibility problems. I'll change that

File: lib/install/config/babel.config.js [hunk 3]
```diff
@@ -30,6 +30,7 @@ module.exports = function(api) {
         {
           forceAllTransforms: true,
           useBuiltIns: 'entry',
+          corejs: 3,
           modules: false,
           exclude: ['transform-typeof-symbol']
         }
```
Comment (janklimo) [inline lib/install/config/babel.config.js]: @somebody32 [This](https://babeljs.io/docs/en/babel-plugin-transform-runtime#corejs) suggests that `@babel/plugin-transform-runtime` (line 56) can specify to use `corejs: 3` as dependency. Were you able to make that work? I'm still seeing errors like these when I make that change:

Screen Shot 2019-04-18 at 11 11 44 PM

File: lib/install/config/babel.config.js [hunk 4]
```diff
@@ -30,6 +30,7 @@ module.exports = function(api) {
         {
           forceAllTransforms: true,
           useBuiltIns: 'entry',
```
Comment (janklimo) [inline lib/install/config/babel.config.js]: @gauravtiwari do you think `usage` should be the new default? From [here](https://github.com/zloirock/core-js/blob/master/docs/2019-03-19-core-js-3-babel-and-a-look-into-the-future.md):

>Until Babel 7.3, useBuiltIns: usage was unstable and not fully reliable: many polyfills were not included, and many others were added without their required dependencies. In Babel 7.4, I tried to make it understand every possible usage pattern.

it does look like it's stable now. If I understand it correctly, using `usage` won't require additional import. When I use:

```js
import "core-js";
```

in my entry point, I get the following message:

```
When setting `useBuiltIns: 'usage'`, polyfills are automatically imported when needed.
Please remove the direct import of `core-js` or use `useBuiltIns: 'entry'` instead.
```

File: lib/install/config/babel.config.js (hunk 3 above)
Comment (somebody32) [reply lib/install/config/babel.config.js]: great point, looks like I need to extend that too (was not using it in our project, that is why didn't spot the problem): https://github.com/zloirock/core-js/blob/master/docs/2019-03-19-core-js-3-babel-and-a-look-into-the-future.md#babelruntime

tldr: try to add `@babel/runtime-corejs3` package

File: lib/install/config/babel.config.js (hunk 4 above)
Comment (somebody32) [reply lib/install/config/babel.config.js]: I would question that. `useBuiltIns: 'usage'` is not assuming anything about user's code, just expect to change `import core-js` to required imports based on supported env, so you need to opt-in in into it by adding that import into one of your packs. 
If we switch that to `usage`, it stops being opt-in and starts automatically adding imports to the code. And from my experience, it still has some edge-cases (because core-js detects relevant import by statically analyzing, so it could not predict what you could have in a runtime), so that creates cases when you still need to add some polyfills by hand.

To summarize: I think `useBuiltIns: 'usage'` is a safer choice, in general, that is very easy to change if you know what you're doing.

Comment (janklimo) [reply lib/install/config/babel.config.js]: Makes sense 👍

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (somebody32) [reply package.json]: Done!

**Comment by somebody32** [comment]
To be honest, I'm not sure what happened to the specs, looks like a configuration issue, and I believe master branch is failing too

File: docs/es6.md [hunk 5]
```diff
@@ -16,13 +16,14 @@ Following ES6/7 features are supported out of the box:
 * Dynamic import() - useful for route level code-splitting
 * Class Fields and Static Properties.
 
-We have also included [babel polyfill](https://babeljs.io/docs/usage/polyfill/)
-that includes a custom regenerator runtime and core-js.
+We have also included [core-js](https://github.com/zloirock/core-js) to polyfill features in the
+older browsers.
 
-Don't forget to import `@babel/polyfill` in your main entry point like so:
+Don't forget add these lines into your main entry point:
```
Comment (maschwenk) [inline docs/es6.md]:
```suggestion
Don't forget to add these lines into your main entry point:
```

Comment (somebody32) [reply docs/es6.md]: Done, thanks!

**Comment by PikachuEXE** [comment]
Any more action is required for this PR?

**Comment by somebody32** [comment]
@PikachuEXE reviewing and merging, I believe

**Comment by gauravtiwari** [comment]
Thanks everyone, will review and merge this over the weekend and make a new release.

**Comment by jaredbeck** [comment]
A recent version of babel (perhaps 7.4.5?) now produces a warning about explicitly setting the `core-js` version.

> WARNING: We noticed you're using the `useBuiltIns` option without declaring a core-js version. Currently, we assume version 2.x when no version is passed. Since this default version will likely change in future versions of Babel, we recommend explicitly setting the core-js version you are using via the `corejs` option...

For now, I'm using the following patch ..

```diff
       (isProductionEnv || isDevelopmentEnv) && [
         require('@babel/preset-env').default,
         {
           forceAllTransforms: true,
           useBuiltIns: 'entry',
+          corejs: '2',
```

Does this PR update need to update the `webpacker:install` rake task accordingly? Sorry if it does already and I'm just not seeing it.

**Comment by somebody32** [comment]
@jaredbeck it is here: https://github.com/rails/webpacker/pull/2031/files#diff-13f93725bce9fb269a61b40f3f4f10deR33

**Comment by PikachuEXE** [comment]
@gauravtiwari Any plan to review this PR soon?

**Comment by jakeNiemiec** [comment]
Made a PR where I added the upgrade instructions to the changelog, feedback appreciated: https://github.com/rails/webpacker/pull/2107

Are there any other outstanding issues with `core-js + babel`?

**Comment by BobbyMcWho** [comment]
This change breaks any project using tailwindcss just as an FYI. 
```
$ bin/webpack-dev-server
ℹ ｢wds｣: Project is running at http://localhost:3035/
ℹ ｢wds｣: webpack output is served from /packs/
ℹ ｢wds｣: Content not from webpack is served from /mnt/d/_repos/test-tailwinds-css/public/packs
ℹ ｢wds｣: 404s will fallback to /index.html
✖ ｢wdm｣: Hash: c31b9d5f78833e23efd1
Version: webpack 4.32.2
Time: 4254ms
Built at: 05/29/2019 9:54:57 AM
                                     Asset       Size       Chunks             Chunk Names
    js/application-d806edd0e75361e03489.js    646 KiB  application  [emitted]  application
js/application-d806edd0e75361e03489.js.map    563 KiB  application  [emitted]  application
                             manifest.json  364 bytes               [emitted]

ERROR in ./app/javascript/css/application.css
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
TypeError: __webpack_require__(...) is not a function
    at Module../node_modules/webpack/buildin/harmony-module.js (/mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:3369:139)
    at __webpack_require__ (/mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:21:30)
    at Module../node_modules/css-loader/dist/runtime/api.js (/mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:3279:41)
    at __webpack_require__ (/mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:21:30)
    at Object../node_modules/css-loader/dist/cjs.js?!./node_modules/postcss-loader/src/index.js?!./app/javascript/css/application.css (/mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:3161:28)
    at __webpack_require__ (/mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:21:30)
    at /mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:85:18
    at Object.<anonymous> (/mnt/d/_repos/test-tailwinds-css/node_modules/css-loader/dist/cjs.js??ref--6-1!/mnt/d/_repos/test-tailwinds-css/node_modules/postcss-loader/src/index.js??ref--6-2!/mnt/d/_repos/test-tailwinds-css/app/javascript/css/application.css:88:10)
    at Module._compile (module.js:652:30)
    at exec (/mnt/d/_repos/test-tailwinds-css/node_modules/mini-css-extract-plugin/dist/loader.js:58:10)
ℹ ｢wdm｣: Failed to compile.
```

For anyone else who manages to find their way here, you have to remove the `corejs: 3` line from your babel.config.js to get successful asset compilation again.

**Comment by jakeNiemiec** [comment]
@BobbyMcWho Thanks for pointing this out. Did you remember to run `bundle exec rails webpacker:install` since this added new packages? [Did you set up webpacker to use tailwind with postcss](https://tailwindcss.com/docs/installation/#using-tailwind-with-postcss)?

**Comment by BobbyMcWho** [comment]
Yeah, I double checked all the config files, the only thing I could get to fix it was to remove the `corejs: 3` line from the `babel.config.js`, and then it compiled just fine afterward.

I can try and put together a reference app when I have a bit of spare time.

**Comment by jakeNiemiec** [comment]
Possibly related:
- https://github.com/babel/babel/issues/10036
- https://github.com/babel/babel/issues/9734

@BobbyMcWho Can you:
- post as much of the error log as you can
- try `corejs: 2` in your config. Babel says [`This gives you the ability to load any version you want, and to update those two packages independently.`](https://babeljs.io/blog/2019/03/19/7.4.0#migration-from-core-js-2)
- does it still happen when **not** using `webpack-dev-server`?

**Comment by BobbyMcWho** [comment]
So the issue happens when the assets are compiled using webpack as part of `bundle exec rails s` on page load, it just displays in the browser console instead. I have figured out that the only place that I need to set `corejs: 2` is in the following section of the babel.config.js:
```
      [
        require('@babel/plugin-transform-runtime').default,
        {
          helpers: false,
          regenerator: true,
          corejs: 2
        }
      ],
```

This lets tailwindcss styles get compiled just fine.

**Comment by jakeNiemiec** [comment]
I have created a PR to fix this here (works with corejs: 3): #2110 

The answer to the problem was that things break when you transpile anything that has to do with transpiling. Picture your modules doing something like this (but poorly):

image

**Comment by gauravtiwari** [comment]
@jakeNiemiec Perhaps, we remove the corejs option in runtime plugin. Sort of confusing that we have env preset that should take care of everything with corejs enabled but we have similar options available for other plugins. Did you find any difference, if we remove corejs option from runtime plugin?

**Comment by jakeNiemiec** [comment]
> Perhaps, we remove the corejs option in runtime plugin.

Fully agree, [plus `create-react-app` followed that same path](https://github.com/facebook/create-react-app/commit/c989b70b89e8945ab4c70a9f3d39e1205bd1007f#diff-7482288d6d13860f434d0b8171f20431R115). 

> Did you find any difference, if we remove corejs option from runtime plugin?

Setting it to false seems to fix things for quite a few users. I made a PR that implements and documents this change: https://github.com/rails/webpacker/pull/2116
